### ISSUE TITLE
[question] How to handle mutation?

### ISSUE THREAD

**Issue by oarriaga** [kimhc6028/pathnet-pytorch#2]
Hello @kimhc6028 thank you for putting forward your code, it has been of great help.
I am currently having trouble understanding two things regarding the code:

1. In the original paper they mention:
 > Mutation takes place with equal probability 1/(3 × 3) per element of the genotype, adding a new random integer from range[−2, 2]

However, I believe they don't mention how to deal whenever the values are below 1 or below 10 (following their encoding of [1, 10]). It seems that you overcome this issue by adding or substracting M

```
    def apply_mutation(self, i, j):
        gene = self.genotype[i][j] + random.randint(-2, 2)
        if gene < 0:
            gene += self.M
        elif gene > self.M - 1:
            gene -= self.M
        self.genotype[i][j] = gene

```
And I just would like to ask how did you come to this conclusion?

2. In the original paper they also mention:

> A maximum of 3 of these units per layer can be included in a single pathway. The genotype describing this pathway is a 3×3 matrix of integers in the range [1, 10].

However, I can't seem to follow where do you handle; or even how the authors handle the encoding whenever there less than 3 active paths, for example 2 or 1 paths. 

Thank you very much.

**Comment by kimhc6028** [comment]
Hello,
1. In the paper, author mentioned 
> A local neighbourhood was used to promote spatial localization of network functionality.

in the last sentence of the first paragraph in 2.2. I could not figure out its meaning clearly, but I assumed that the edge nodes (1st node and 10th node) have to be neighbourhood. 
I was worried that if I don't subtract M at the end of nodes, many genotypes will be converged to the 1st and 10th nodes.
However I am not quite sure if this is really needed. It would be interesting if it works without this :)

2. Yes, this is one part that I was confused. My conclusion is, we don't need to handle this. If you see README in this repo, you can find there is only two nodes are activated in first and second layer after Generation 100 (fixed path). Like this, paths will be overlapped (i.e., [[9,10,10],[1,1,8],[2,7,8]]). Of course I am worried if doubling the output of one path will affect the result, but I couldn't find any clue for this in the paper too

I appreciate your feedback. Thank you so much.

**Comment by oarriaga** [comment]
Hi @kimhc6028,
Thank you for your fast response.
1. a) Yes, exactly. They want to mutate the paths such that the mutated path is close by the original path. In this case it leaves open how to deal when mutating path module 1 and 10. As you mentioned, connecting 1 and 10 could be a possibility. Or also limiting the arguments from which the modules 1 and 10 can sample e.g. all modules 1 can only sample arguments from [0, 2] and all modules 10 can sample from [-2, 0].
b) I don't think that this "local" bias could deteriorate the performance, but still it should be tested.

2. Thank you, yes this is also another important issue. Please correct me if I am wrong, you are then doubling the path which means that the reduce sum modules will sum it up twice? Again, the authors do not mention anything regarding this. However, what I have been thinking is that the 0 argument might play a role in their implementation. For example they mention that the initial genotypes are encoded using values between [1, 9] (they don't take zeros) however, the mutation operator seems that it can change the value to 0 since they are sampling from integers values between [-2, 2], which could be that the zero encodes the no-path. I think your implementation uses the integer 0 to encode the first module of each layer.  What do you think?

Thanks Octavio

**Comment by oarriaga** [comment]
Looking at the mutation function of the [pathnet tensorflow implementation](https://github.com/jaesik817/pathnet/blob/master/pathnet.py) it seems that they are clipping the values less than zero and bigger than 10

```
def mutation(geopath,L,M,N):
  for i in range(L):
    for j in range(M):
      if(geopath[i,j]==1):
        rand_value=int(np.random.rand()*L*N);
        if(rand_value<=1):
          geopath[i,j]=0;
          rand_value2=np.random.randint(-2,2);
          rand_value2=rand_value2-2;
          if(((j+rand_value2)>=0)&((j+rand_value2)<M)):
            geopath[i,j+rand_value2]=1;
          else:
            if((j+rand_value2)<0):
              geopath[i,0]=1;
            elif((j+rand_value2)>=M):
              geopath[i,M-1]=1;
  return geopath;
```

I think they are using a one-hot encoding and whenever their argument is less than zero they take the first module (represented with argument 0)
```
             if((j+rand_value2)<0):
              geopath[i,0]=1;
```
And when the argument is bigger than the last possible path module the only put a path in the last possible module

```
            elif((j+rand_value2)>=M):
              geopath[i,M-1]=1;
```

**Comment by jsikyoon** [comment]
Hi, Guys

Good Morning :) (in Korea)

I got a msg from kimhc6028 for this issue.

I also confused the point when I first implemented the PathNet.

In during of discussing with the author of the paper, I checked original implementation is similar to our ones.

However, I did check kimhc6028 mutation method (over limitation case, cycling the index with M) and author mutation method. In my checking, there are not lots of difference of performance in both method.

Please just considering those things :)

The bellowed ones are got from the author to compare implementations.

  DAG = True: 

    def mutate(self, path):
        if DAG: 
          for i in xrange(self.genotypeLength):
              for p in xrange(FLAGS.MAX_DAG_WIDTH): 
                if random.uniform(0,1) < FLAGS.mutation_rate:  
                    path[i][p] =  path[i][p] + random.randint(-2,2)
                    if path[i][p] < 0: 
                      path[i][p] = 0
                    if path[i][p] > FLAGS.network_width-1: 
                      path[i][p] = FLAGS.network_width-1
        else: 
           for i in xrange(self.genotypeLength):
            if random.uniform(0,1) < FLAGS.mutation_rate:
              r = random.randint(0,FLAGS.network_width-1)
              for p in xrange(FLAGS.MAX_DAG_WIDTH):   
                path[i][p] = r 

        return path


BRs, Jaesik.

**Comment by kimhc6028** [comment]
Sorry for late. 
Okay, looks like we solved first issue? :)
For the second issue, yes you are right, reduce sum module will add same output twice.
In the paper, all 10 modules are connected to next layer (see p.7). So, at least the authors do not used no-path for 0th module. Still, the author mentioned that they used simplest evolution algorithm, which implies that we can test many other versions of structures. Dropping out unnecessary module would be interesting :), especially if we reward for computational efficiency!

**Comment by oarriaga** [comment]
Hello guys,
Thank you @kimhc6028  and @jaesik817 for your fast responses.
Yes it seems that the first issue is then solved and the original implementation clips the arguments when they are less or bigger than the possible module.

I was wrong regarding arg 0 being the no-path since the values of the arguments are getting summed and not replaced as I was thinking. @jaesik817  do you have any idea of whether or not original authors sum the path twice? or how are you/the authors implement the no-path encoding?

Thank you guys for your awesome input. 
regards from Germany.

**Comment by jsikyoon** [comment]
Hi, @kimhc6028 , @oarriaga ,

First, I checked the part on original implementation. In that, just one-hot encoding is used for twice selected module (The original one used if statement of tensorflow to decide the module index is included in path matrix or not).

My implementation is also just one-hot. (My implementation is just to multiply 0 or 1 to output of module, if index of which is included in path then multiply 1. In another case, my ones multiply 0. **Because that is faster than original implementation.**)

I also think they did not use no-path encoding (I also did not use those concept.).

I did not consider twice summation, because, for me, the index of geotype indicates the opened/closed "path", not concept of summation. I think considering multiple selection on same module can be interesting try :).

BRs, Jaesik.

**Comment by kimhc6028** [comment]
According to @jaesik817, he summed up all activated modules only once, unlike my implementation which sums up all the overlapped paths. Am I right?

**Comment by oarriaga** [comment]
Thank you @jaesik817 ,
Then everything is semi-clear now.

Please correct me if I am wrong, you are using one-hot encoding for the genotypes where 1 means "we will use this module in this layer for our path" and zero means no connection. If so, the only issue I see is that the authors specifically mention:

> The genotype describing this pathway is a 3×3 matrix of integers in the range [1, 10].

But for me the one-hot encoding seems more appropriate than the range [1, 10]. But yeah, I guess if you @jaesik817 saw the original code and they do the one-hot then I guess we are in the correct path (pun intended ;-) )

regards,
Octavio

**Comment by jsikyoon** [comment]
@oarriaga and @kimhc6028 

Haha 

They are using one-hot by using conditional statement of tensorflow as above my comment.

My thinking for difference between mine and the original is as followed.

**The functionality between two implementation is same.**

My implementation uses L by M boolean matrix for checking the path, and
the paper uses L by N integer matrix for same purpose.

The paper/My implementation check that j-th module of i-th layer is included in the path or not 
by checking j is included in  i-th row, and
by checking the element in i-th row and j-th column is 1, respectively.

If you have any more questions, then please feel free to ask that :)

BRs,
Jaesik.

**Comment by jsikyoon** [comment]
@oarriaga 

One more thing to note that

the pathnet structure and the rule for using "optimal path of previous task " on next tasks of my implementation are different to the paper.

Please read readme :-) or ask to me :)

Thanks.

**Comment by oarriaga** [comment]
Perfect, I think everything is clear, I guess I will just have to implement this part and see if I stumble again into any doubts. I will look carefully into your tensorflow implementation and contact you if I have any. 

Thank you guys @jaesik817, @kimhc6028  for your very fast-responses and willingness :D I will let you know about my results.
Regards
